src/frontend/state/InjectedProviderHandler.ts

```diff
@@ -0,0 +1,159 @@
+      await this.waitForConnection()
+      /* eslint-disable-next-line */
+      //@ts-ignore
+      const value = await window.ethereum?.send(...args)
+      window.api.returnExtensionRequest(id, value)
+    } catch (err) {
+      console.error(`error during send: ${err}`)
+      window.api.errorExtensionRequest(id, err)
+    }
+  }
+
+  async handleSendAsync(
+    event: Electron.IpcRendererEvent,
+    id: number,
+    args: unknown[]
+  ) {
+    try {
+      await this.waitForConnection()
+      /* eslint-disable-next-line */
+      //@ts-ignore
+      const value = await window.ethereum?.sendAsync(...args)
+      window.api.returnExtensionRequest(id, value)
+    } catch (err) {
+      console.error(`error during send: ${err}`)
+      window.api.errorExtensionRequest(id, err)
+    }
+  }
+
+  bindEthereumListeners() {
+    window.addEventListener('message', (event: MessageEvent) => {
+      /**
+       * An {"isTrusted": true} message is sent every few seconds and spams the logs.
+       * This makes debugging difficult in Sentry so we filter these out.
+       */
+      if (
+        event !== undefined &&
+        Object.hasOwn(event, 'isTrusted') &&
+        Object.keys(event).length === 1
+      ) {
+        return
+      }
+    })
+
+    /* eslint-disable-next-line */
+    //@ts-ignore
+    window.ethereum?.on('accountsChanged', (accounts: string[]) => {
+      window.api.extensionOnEvent('accountsChanged', accounts)
+    })
+
+    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
+    window.ethereum?.on('disconnect', (error: any) => {
+      // Relevant issue https://github.com/MetaMask/metamask-extension/issues/13375
+      if (error.code === 1013) {
+        console.log('MetaMask disconnected from chain. Reconnecting...')
+        return
+      }
+      window.api.extensionOnEvent('disconnect', error)
+    })
+
+    /* eslint-disable-next-line */
+    //@ts-ignore
+    window.ethereum?.on('connect', (connectInfo: string) => {
+      window.api.extensionOnEvent('connect', connectInfo)
+    })
+
+    /* eslint-disable-next-line */
+    //@ts-ignore
+    window.ethereum?.on('chainChanged', (chainId: number) => {
+      window.api.extensionOnEvent('chainChanged', chainId)
+    })
+
+    const removeRequestListener = window.api.handleMetamaskExtensionRequests(
+      this.handleRequest.bind(this)
+    )
+    const removeSendListener = window.api.handleMetamaskExtensionSends(
+      this.handleSend.bind(this)
+    )
+    const removeSendAsyncListener =
+      window.api.handleMetamaskExtensionSendAsyncs(
+        this.handleSendAsync.bind(this)
+      )
+
+    this.ethereumListenersBound = true
+
+    return () => {
+      removeRequestListener()
+      removeSendListener()
+      removeSendAsyncListener()
+    }
+  }
+
+  constructor() {
+    makeAutoObservable(this)
+    const bindListeners = (): boolean => {
+      if (typeof window.ethereum !== 'undefined') {
+        console.log('binding ethereum listeners')
```

Comment:
Why instead of using all those console.logs, that would be complicated to debug if an user has some issues, you don't call `window.api.logInfo` (logError, etc)? So then it would be included in the app logs that the users would send when opening an issue on zendesk.

Reply:
This is in the browser window, so we need to be able to see the logs in that particular browser window. If we were using this in multiple browser windows and only logging to the main node.js process console, it would be impossible to debug. I can do both though so we can see if at least one of the windows binded listeners to window.ethereum

Reply:
updated